mono/metadata/cominterop.c

```diff
@@ -2660,47 +2695,65 @@ cominterop_ccw_get_ids_of_names_impl (MonoCCWInterface* ccwe, gpointer riid,
 	static MonoClass *ComDispIdAttribute = NULL;
 	ERROR_DECL (error);
 	MonoCustomAttrInfo *cinfo = NULL;
-	int i,ret = MONO_S_OK;
-	MonoMethod* method;
+	int i, j, methodcount, pos, ret = MONO_S_OK;
+	MonoMethod* method = NULL;
 	gchar* methodname;
-	MonoClass *klass = NULL;
+	MonoClass* iface = ccwe->iface;
 	MonoCCW* ccw = ccwe->ccw;
-	MonoObject* object = mono_gchandle_get_target_internal (ccw->gc_handle);
+	MonoObject* object;
 
 	/* Handle DispIdAttribute */
 	if (!ComDispIdAttribute)
 		ComDispIdAttribute = mono_class_load_from_name (mono_defaults.corlib, "System.Runtime.InteropServices", "DispIdAttribute");
 
-	g_assert (object);
-	klass = mono_object_class (object);
-
 	if (!mono_domain_get ())
 		 mono_thread_attach (mono_get_root_domain ());
 
+	if (iface == mono_class_get_idispatch_class ()) {
+		object = mono_gchandle_get_target_internal (ccw->gc_handle);
+		g_assert (object);
+		iface = mono_object_class (object);
+	}
+
+	mono_class_setup_methods (iface);
+	methodcount = mono_class_get_method_count (iface);
+
 	for (i=0; i < cNames; i++) {
 		methodname = mono_unicode_to_external (rgszNames[i]);
 
-		method = mono_class_get_method_from_name_checked(klass, methodname, -1, 0, error);
-		if (method && is_ok (error)) {
+		if (mono_class_is_interface (iface))
+			pos = 0;
+		else
+			/* TODO: methods on Object should be properly included? */
+			pos = 4;
+
+		for (j = 0; j < methodcount; j++) {
+			method = m_class_get_methods (iface) [j];
+
+			if (strcmp(method->name, methodname) == 0)
+				break;
+
+			pos++;
+		}
+
+		if (j < methodcount) {
 			cinfo = mono_custom_attrs_from_method_checked (method, error);
 			mono_error_assert_ok (error); /* FIXME what's reasonable to do here */
 			if (cinfo) {
-				MonoObject *result = mono_custom_attrs_get_attr_checked (cinfo, ComDispIdAttribute, error);
+				MonoDispIdAttribute *result = (MonoDispIdAttribute*)mono_custom_attrs_get_attr_checked (cinfo, ComDispIdAttribute, error);
 				mono_error_assert_ok (error); /*FIXME proper error handling*/;
 
 				if (result)
-					rgDispId[i] = *(gint32*)mono_object_unbox_internal (result);
+					rgDispId[i] = result->dispid;
 				else
-					rgDispId[i] = (gint32)method->token;
+					rgDispId[i] = 0x60020000 + pos;
```

Comment:
What does this value mean?

Reply:
I don't know, it was found by testing. This value is commonly used in IDispatch definitions in Windows headers for sequential method id's, in a similar way to what .NET is doing, but I couldn't find a define or documentation for it.